language/move-lang/src/typing/translate.rs

```diff
@@ -571,11 +572,118 @@ fn struct_def(context: &mut Context, s: &mut N::StructDefinition) {
+            context,
+            &type_params,
+            &idx_ty.1,
+            Some(NonPhantomPos::FieldType),
+        );
+    }
+}
+
+enum NonPhantomPos {
+    FieldType,
+    TypeArg,
+    Other,
+}
+
+fn check_phantom_params(
+    context: &mut Context,
+    type_parameters: &BTreeMap<TParamID, (bool, Loc)>,
+    ty: &Type,
+    non_phantom_pos: Option<NonPhantomPos>,
+) {
+    use NonPhantomPos::*;
+
+    match &ty.value {
+        Type_::Param(param) => {
+            // Don't bother checking unbound type parameters as it should fail somewhere else
+            match (type_parameters.get(&param.id), non_phantom_pos) {
+                (Some(&(is_phantom, param_decl_loc)), Some(non_phantom_pos)) if is_phantom => {
+                    invalid_phantom_use_error(
+                        context,
+                        non_phantom_pos,
+                        param_decl_loc,
+                        param,
+                        ty.loc,
+                    );
+                }
+                _ => {}
+            }
+        }
+        Type_::Apply(_, n, ty_args) => match &n.value {
+            TypeName_::Builtin(_) => {
+                for ty_arg in ty_args {
+                    check_phantom_params(context, type_parameters, ty_arg, Some(TypeArg));
+                }
+            }
+            TypeName_::ModuleType(m, n) => {
+                let param_is_phantom: Vec<_> = context
+                    .struct_tparams(m, n)
+                    .iter()
+                    .map(|param| param.is_phantom)
+                    .collect();
+                // Length of params and args may be different but we can still report errors
+                // for parameters with information
+                for (is_phantom, ty_arg) in param_is_phantom.into_iter().zip(ty_args) {
+                    check_phantom_params(
+                        context,
+                        type_parameters,
+                        ty_arg,
+                        (!is_phantom).then(|| TypeArg),
+                    );
+                }
+            }
+            // Tuples cannot appear in structs, but we still report them as a non-phantom position in
+            // case they are ever allowed.
+            TypeName_::Multiple(_) => {
+                for ty_arg in ty_args {
+                    check_phantom_params(context, type_parameters, ty_arg, Some(Other));
+                }
+            }
+        },
+        // References cannot appear in structs, but we still report them as a non-phantom position
+        // in case they are ever allowed.
+        Type_::Ref(_, inner) => check_phantom_params(context, type_parameters, &inner, Some(Other)),
+        Type_::Var(_) | Type_::Anything | Type_::Unit | Type_::UnresolvedError => {}
     }
 }
 
+fn invalid_phantom_use_error(
+    context: &mut Context,
+    non_phantom_pos: NonPhantomPos,
+    param_decl_loc: Loc,
+    param: &TParam,
+    ty_loc: Loc,
+) {
+    use NonPhantomPos::*;
```

Comment:
I used to do this a lot more, but have sort of stopped once I realized Rust does a bad job of dealing with variants vs new bindings. So I will leave using this here up to you